CRM/ManualDirectDebit/Hook/PostProcess/Contribution/DirectDebitMandate.php

```diff
@@ -91,7 +41,9 @@ private function createMandate() {
     $queryResult->fetch();
     $generatedMandateId = $queryResult->id;
 
-    return $generatedMandateId;
+    // sets mandate id, for saving dependency between mandate and contribution
+    $mandateContributionConnector = CRM_ManualDirectDebit_Hook_MandateContributionConnector::getInstance();
+    $mandateContributionConnector->setMandateId($generatedMandateId);
```

Comment:
A little worried about how `$generatedId` is being calculated, as it may be subject to race conditions. We could embed the whole thing in a transaction tough, to avoid weird problems.

Eg.
```php
$tx = new CRM_Core_Transaction();
try {
  $sqlInsertedInDirectDebitMandate = "INSERT INTO `$tableName` (`entity_id`) VALUES (%1)";
  CRM_Core_DAO::executeQuery($sqlInsertedInDirectDebitMandate, [
    1 => [
      $this->form->getVar('_contactID'),
      'String',
    ],
  ]);

  $sqlSelectedDebitMandateID = "SELECT MAX(`id`) AS id FROM `$tableName` WHERE `entity_id` = %1";
  $queryResult = CRM_Core_DAO::executeQuery($sqlSelectedDebitMandateID, [
    1 => [
      $this->form->getVar('_contactID'),
      'String',
    ],
  ]);
  $queryResult->fetch();
  $generatedMandateId = $queryResult->id;
} catch (Exception $e) {
  $tx->rollback();
  throw $e;
}
```

There are some other ways transactions can be implemented in CiviCRM, check:
https://docs.civicrm.org/dev/en/latest/framework/database/transactions/

It might be worth to look at embedding larger parts of mandates and direct debits processing within transactions, to improve the robustness and reliability of the extension. Since we're dealing with money, we should do our best not to allow for inconsistencies if an exception should occur amidst processing.

CC @guanhuan